src/NumericalAlgorithms/LinearAlgebra/FindGeneralizedEigenvalues.hpp

```diff
@@ -0,0 +1,51 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function find_generalized_eigenvalues.
+
+#pragma once
+
+/// \cond
+class DataVector;
+class Matrix;
+namespace gsl {
+template <typename T>
+class not_null;
+} // namespace gsl
+/// \endcond
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief Solve the generalized eigenvalue problem for two matrices.
+ *
+ * This function uses the lapack routine dggev to solve the
+ * generalized eigenvalue problem \f$A v_a =\lambda_a B v_a \f$
+ * for the generalized eigenvalues \f$lambda_a\f$ and corresponding
+ * eigenvectors \f$v_a\f$.
+ * `A` and `B` are each a `Matrix`; they correspond to square
+ * matrices \f$A\f$ and \f$B\f$ that are the same order \f$N\f$.
+ * `eigenvalues_real_part` is a `DataVector` of size \f$N\f$ that
+ * will store the real parts of the eigenvalues,
+ * `eigenvalues_imaginary_part` is a `DataVector` of size \f$N\f$
+ * that will store the imaginary parts of the eigenvalues.
+ * Complex eigenvalues always form complex conjugate pairs, and
+ * the \f$j\f$ and \f$j+1\f$ eigenvalues will have the forms
+ * \f$a+ib\f$ and \f$a-ib\f$, respectively. The eigenvectors
+ * are returned as the columns of a square `Matrix` of order $N$
+ * called `eigenvectors`. If eigenvalue j is real, then column j of `v` is
+ * the corresponding eigenvector. If eigenvalue j and j+1 are
+ * complex-conjugate pairs, then the eigenvector for
+ * eigenvalue j is (column j) + i (column j+1), and the
+ * eigenvector for eigenvalue j+1 is (column j) - i (column j+1).
+ * Note: the outputs are passed as gsl::not_null pointers, so when
+ * calling this function, pass in the addresses for those
+ * (e.g. find_generalized_eigenvalues(&real, &imag, &vecs, A, B);).
+ * For testing, set the option `force_lapack_failure` to true to
+ * cause the lapack call, dggev_(), to return nonzero.
+ */
+void find_generalized_eigenvalues(
+    gsl::not_null<DataVector*> eigenvalues_real_part,
+    gsl::not_null<DataVector*> eigenvalues_imaginary_part,
+    gsl::not_null<Matrix*> eigenvectors, const Matrix& A,
+    const Matrix& B, bool force_lapack_failure = false) noexcept;
```

Comment:
From the math, I would expect a failure if B=0.

Reply:
Please use some math to make it fail, we don't want an API for testing purposes, that's bad.

Reply:
Yeah, I see what you're saying. But I tried for about an hour different bad matrices, like the ones you suggest, and they end up triggering FPEs instead of triggering the failure I'm testing here (info != 0). 

I don't want to remove the check, because I don't know for sure whether some matrix I can't think of will lead to info !=0. But if I keep it, I have to test it somehow. The only way I know that makes info !=0 deliberately is making a choice for a size parameter outside what lapack expects, which normally never happens because correctly written code won't do this. 

Given this context, do you want me to i) remove the check, or ii) leave the API as-is?

Reply:
I see. In that case I want the non-option iii: leave the `if (UNLIKELY(info != 0))` (note, please add the `UNLIKELY`, this tells the compiler we don't expect the statement to evaluate true), but also remove the bool from the API. We can't test everything, and this check is more of an internal consistency check that you shouldn't generally reach through the interface :)